app/docker/views/images/import/importImageController.js

```diff
@@ -1,35 +1,71 @@
+import { PorImageRegistryModel } from 'Docker/models/porImageRegistry';
+
 angular.module('portainer.docker').controller('ImportImageController', [
   '$scope',
   '$state',
   'ImageService',
   'Notifications',
   'HttpRequestHelper',
-  function ($scope, $state, ImageService, Notifications, HttpRequestHelper) {
+  'Authentication',
+  'ImageHelper',
+  'endpoint',
+  function ($scope, $state, ImageService, Notifications, HttpRequestHelper, Authentication, ImageHelper, endpoint) {
     $scope.state = {
       actionInProgress: false,
     };
 
+    $scope.endpoint = endpoint;
+
+    $scope.isAdmin = Authentication.isAdmin();
+
     $scope.formValues = {
       UploadFile: null,
       NodeName: null,
+      RegistryModel: new PorImageRegistryModel(),
     };
 
-    $scope.uploadImage = function () {
+    $scope.setPullImageValidity = setPullImageValidity;
+    function setPullImageValidity(validity) {
+      $scope.state.pullImageValidity = validity;
+    }
+
+    async function tagImage(id) {
+      const registryModel = $scope.formValues.RegistryModel;
+      if (registryModel.Image) {
+        const image = ImageHelper.createImageConfigForContainer(registryModel);
+        try {
+          await ImageService.tagImage(id, image.fromImage);
+        } catch (err) {
+          Notifications.error('Failure', err, 'Unable to tag image');
+        }
+      }
+    }
+
+    $scope.uploadImage = async function () {
       $scope.state.actionInProgress = true;
 
       var nodeName = $scope.formValues.NodeName;
       HttpRequestHelper.setPortainerAgentTargetHeader(nodeName);
       var file = $scope.formValues.UploadFile;
-      ImageService.uploadImage(file)
-        .then(function success() {
+      try {
+        const { data } = await ImageService.uploadImage(file);
+        if (data.error) {
+          Notifications.error('Failure', data.error, 'Unable to upload image');
+        } else if (data.stream) {
+          var regex = /Loaded.*?: (.*?)\n$/g;
+          var imageIds = regex.exec(data.stream);
+          if (imageIds && imageIds.length == 2) {
+            await tagImage(imageIds[1]);
+            $state.go('docker.images.image', { id: imageIds[1] }, { reload: true });
+          }
           Notifications.success('Images successfully uploaded');
-        })
-        .catch(function error(err) {
-          Notifications.error('Failure', err, 'Unable to upload image');
-        })
-        .finally(function final() {
-          $scope.state.actionInProgress = false;
-        });
+        } else {
+          Notifications.success('The uploaded tar file contained multiple images. The provided tag therefore has been ignored.');
+        }
+        $scope.state.actionInProgress = false;
```

Comment:
this should happen in finally, like in the old code (otherwise spinner will keep going on failure)

Reply:
```
try {
} catch(e){
} finally {
$scope.state.actionInProgress = false;
}
```

Reply:
done